pubsub/envelope.go

```diff
@@ -30,29 +33,47 @@ type CloudEventsEnvelope struct {
 	Subject         string      `json:"subject"`
 }
 
-// NewCloudEventsEnvelope returns a new CloudEventsEnvelope
+// NewCloudEventsEnvelope returns CloudEventsEnvelope from data or a new one when data content was not
+// already an existing CE. Invoked by either the HTTP or gRPC pug handles where:
+//  id = new UUID
+// 	source = handling app ID
+//  eventType = const DefaultCloudEventType from this package ("com.dapr.event.sent")
+//  subject = diagnostics span context W3C as string
 func NewCloudEventsEnvelope(id, source, eventType, subject string, data []byte) *CloudEventsEnvelope {
-	if eventType == "" {
-		eventType = DefaultCloudEventType
+	var ce CloudEventsEnvelope
+	err := jsoniter.Unmarshal(data, &ce)
+	if err == nil && ce.ID != "" {
```

Comment:
Should this case be an error?  I'm reading this as: the user provided a CE but it's an invalid CE because id was not specified.

Reply:
No, this is a valid CloudEvent (no error and it did unmarshal into CloudEvent). There are 3 scenarios this code covers when the user posts content: 

1. content is not JSON (deserialization error), set type to TXT and create a brand new CloudEvents envelop
2. content is valid JSON (no deserialization error) but the unmarshalling didn't match any fields from the CloudEvent so set type to JSON and set fields on the new CloudEvents envelop
3. content is valid JSON in CloudEvent format (no deserialization error), return the unmarshalled struct as is

Reply:
I see, thanks for explaining.  In number 2 there is the possibility that the non-CE json happens to have an "ID" field, which seems like a field a lot of objects may have, which would make this go down the wrong path.  Perhaps using a different field like "SpecVersion" may decrease the chance of this.

Reply:
Good call! I will combine SpecVersion, Type, and Source to be more defensive on this.

Reply:
Come to think of it, I included all of the CloudEvent required fields (ID, SpecVersion, Type, Source, DataContentType). PTAL